src/Polysemy/Internal/TH/Effect.hs

```diff
@@ -75,135 +79,185 @@ makeSem = genFreer True
-genType _       = fail "genSig expects a GADT constructor"
-
--- | Turn all (KindedTV tv StarT) into (PlainTV tv) in the given type
--- This can prevent the need for KindSignatures
-simplifyBndrs :: Maybe Type -> Type -> Type
-simplifyBndrs star = everywhere (mkT $ simplifyBndr star)
-
--- | Turn TvVarBndrs of the form (KindedTV tv StarT) into (PlainTV tv)
--- This can prevent the need for KindSignatures
-simplifyBndr :: Maybe Type -> TyVarBndr -> TyVarBndr
-simplifyBndr (Just star) (KindedTV tv k) | star == k = PlainTV tv
-simplifyBndr _ (KindedTV tv StarT) = PlainTV tv
-simplifyBndr _ bndr = bndr
-
--- | Generates a type signature of the form
--- @x :: Member (Effect e) r => a -> b -> c -> Sem r r@.
-genSig :: Con -> Q Dec
-genSig con = do
-  let
-    getConName (ForallC _ _ c) = getConName c
-    getConName (GadtC [n] _ _) = pure n
-    getConName c = fail $ "failed to get GADT name from " ++ show c
-  conName <- getConName con
-  (t, _, k) <- genType con
-  pure $ SigD (getDeclName conName) $ simplifyBndrs k t
 
+------------------------------------------------------------------------------
 -- | Folds a list of 'Type's into a right-associative arrow 'Type'.
 foldArrows :: [Type] -> Type
-foldArrows = foldr1 (AppT . AppT ArrowT)
+foldArrows = foldr1 $ AppT . AppT ArrowT
+
+------------------------------------------------------------------------------
+-- | Extracts name from type variable (possibly nested in signature and/or
+-- some context), returns 'Nothing' otherwise.
+tvName' :: Type -> Maybe Name
+tvName' = \case
+  ForallT _ _ t -> tvName' t
+  SigT t _      -> tvName' t
+  VarT n        -> Just n
+  ParensT t     -> tvName' t
+  _             -> Nothing
+
+------------------------------------------------------------------------------
+-- | Converts equality constraint with type variable to name and type pair or
+-- leaves predicate as is
+eqPairOrCxt :: Pred -> Either (Name, Type) Pred
+eqPairOrCxt p = case asEqualPred p of
+  Just (VarT n, b) -> Left (n, b)
+  Just (a, VarT n) -> Left (n, a)
+  _                -> Right p
+
+------------------------------------------------------------------------------
+-- | Constructs name from base of input name with function applied
+mapNameBase :: (String -> String) -> Name -> Name
+mapNameBase f = mkName . f . nameBase
+
+------------------------------------------------------------------------------
+-- TODO: separate into some module for general utilities (or find them in
+-- available dependencies)
+
+partitionWith :: (a -> Either b c) -> [a] -> ([b], [c])
+partitionWith f = foldl' choose ([], []) where
+  choose = flip $ either (first . (:)) (second . (:)) . f
 
+splitFromEnd :: Int -> [a] -> ([a], [a])
+splitFromEnd n lst = go lst $ drop n lst where
+  go xs     []     = ([], xs)
+  go (x:xs) (_:ys) = first (x:) $ go xs ys
 
-freeVars :: Data a => a -> [Name]
-freeVars = mkQ [] $ \case
-  VarT n -> [n]
-  _ -> []
+  go []     (_:_)  = undefined
```

Comment:
yucky -- why is this `undefined`?

Reply:
`-Wincomplete-patterns`...

Reply:
It's a bad idea to disable it, so I'm being explicit instead - this is not Idris, so I can't just prove this doesn't happen...

Reply:
I meant more generally :) like why _can't_ this case happen? Maybe add a comment with your reasoning. Change the `undefined` to an error with the name of the function as well, so if it _does_ go wrong we have a chance of finding it.

Reply:
Added descriptive `error`.